### PR TITLE
clarify correctness of `black_box`

### LINKED ISSUE #107957: black_box docs correctness claim incorrect
### Location

https://doc.rust-lang.org/stable/std/hint/fn.black_box.html

### Summary

The documentation of `black_box` states: 

> Programs cannot rely on black_box for correctness in any way.

but this is not accurate since programs can rely on it behaving like the identity function.
It would be better to reword this as 

> Programs cannot rely on black_box for correctness beyond it behaving as the identity function.

or similar, e.g., 

> Programs cannot rely on the optimisation-inhibiting effects of black_box for correctness.

### PR DESCRIPTION
**Pull request by callumio**
Resolves #107957, as described in the related issue.

No code changes, just documentation.

### CODE REVIEW

**Comment by cuviper** [comment]
r? libs-api

**Review by dtolnay — APPROVED** [review]
Thanks!
